### PR TITLE
[NFC, Incremental] Refactoring to facilitate replacing the inputDependencySourceMap with the right thing

### PR DESCRIPTION
**Pull request by davidungar**
A refactoring of `inputDependencySourceMap` prior to really fixing the below:

See https://github.com/apple/swift-driver/pull/662

> There is a remaining lurking bug here that involves an extra step.
Suppose you build in a normal way with foo.swift/foo.c setup. Then you
rebuild and cancel it after a frontend has emitted a new swiftdeps file,
finally, you crash the driver by exploiting this bug. The driver is now
in a state where swiftdeps disagree with the priors which disagree with
the output file map.

> All this to say, we should take the earliest opportunity to remove
BidirectionalMap and its uses. But this commit is not the one to make
such a structural change to the driver.

Start on that.

rdar://77998890

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test macOS Platform

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// MARK: - Populating
+extension InputDependencySourceMap {
+  public enum AdditionPurpose {
+    case mocking,
+         buildingFromSwiftDeps,
+         readingPriors,
+         inputsAddedSincePriors }
+  @_spi(Testing) public mutating func addEntry(_ input: TypedVirtualPath,
+                                               _ dependencySource: DependencySource,
+                                               `for` _ : AdditionPurpose) {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:60]: Nit: You don't need to quote keywords when they're external labels. If Xcode is highlighting this incorrectly, please file a radar.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    _ eachFn: (TypedVirtualPath, DependencySource) -> Void
+  ) {
+    biMap.forEach(eachFn)
+  }
+}
+
+// MARK: - Populating
+extension InputDependencySourceMap {
+  public enum AdditionPurpose {
+    case mocking,
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:54]: Please document these. Their names alone don't reveal much at the point of usage.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 3]
```diff
@@ -219,17 +204,47 @@ extension ModuleDependencyGraph {
     return nodeFinder.findNodes(for: source).map {!$0.isEmpty}
       ?? false
   }
-
-  /// Return true on success
-  func populateInputDependencySourceMap() -> Bool {
+  
+  /// Returns: false on error
+  func populateInputDependencySourceMap(
+    `for` purpose: InputDependencySourceMap.AdditionPurpose
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:210]: Nit: Same here. Don't need the quotes around `for`.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+extension InputDependencySourceMap {
+  @_spi(Testing) public func getSourceIfKnown(for input: TypedVirtualPath) -> DependencySource? {
+    biMap[input]
+  }
+
+  @_spi(Testing) public func getInputIfKnown(for source: DependencySource) -> TypedVirtualPath? {
+    biMap[source]
+  }
+
+  @_spi(Testing) public func enumerateToSerializePriors(
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:44]: This name conflates the API's purpose with its usage. I think the swift-ier name is something closer to just `enumerate(_:)`.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  // and a regular map from dependency sources to swift files -
+  // since that direction really is one-to-one.
+  
+  public typealias BiMap = BidirectionalMap<TypedVirtualPath, DependencySource>
+  @_spi(Testing) public var biMap = BiMap()
+}
+
+// MARK: - Accessing
+extension InputDependencySourceMap {
+  @_spi(Testing) public func getSourceIfKnown(for input: TypedVirtualPath) -> DependencySource? {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:36]: Just `source(for:)` would do here.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    biMap[input]
+  }
+
+  @_spi(Testing) public func getInputIfKnown(for source: DependencySource) -> TypedVirtualPath? {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:40]: just `input(for:)` as well

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 1 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:60]: Thanks, will do.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 2 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:54]: Okey-doke!

**Review by CodaFi — CHANGES_REQUESTED** [review]
Some documentation of this new interface would be welcome. API notes are in-line.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 4 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:44]: The conflation is quite deliberate here. The enclosing type is meant to operate at the level of *intention*. This level distinction paves the way for the greater structural change to come.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 5 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:36]: Yes, it would. And I'll remove the "get" prefix.

 The idea here was to draw a very explicit distinction with `ModuleDependencyGraph.getRequiredSource`. In an earlier draft, both were named `getSource` and it was not easy to search for one as opposed to the other, because the only difference in usage was the result type. So, I decided to rename at least one, for the sake of clarity. Then, for the sake of symmetry (and clarity), I renamed them both.
So, I would rather keep the "IfKnown" part.

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:36]: Hmm... after eliminating the "get"s, the "IfKnown" etc., moves to the keyword. You'll see after I push. Does that look better?

**Comment by davidungar** [comment]
> Some documentation of this new interface would be welcome. API notes are in-line.

Yes, I'll add.

**Comment by davidungar** [comment]
Thanks for the comments; they helped! Have addressed them, and now will fix the conflicts.

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
Thanks, @CodaFi , for productive comments. I've made the changes requested, modulo keeping some intentionality in the function names. This rev may not be there yet, so please let me know what you see.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 4 above)
Comment (CodaFi) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:58]: I understand that, but Swift does not use the name of the API in the same way that Objective-C or Smalltalk might encourage the use of intention-revealing selectors.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  /// - Returns: the corresponding `DependencySource`, or nil if none known.
+  @_spi(Testing) public func sourceIfKnown(for input: TypedVirtualPath) -> DependencySource? {
+    biMap[input]
+  }
+
+  /// Find where the source file is for a given swiftdeps file.
+  ///
+  /// - Parameter source: A `DependencySource` (containing a swiftdeps file)
+  /// - Returns: the corresponding input source file, or nil if none known.
+  @_spi(Testing) public func input(ifKnownFor source: DependencySource) -> TypedVirtualPath? {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:51]: `ifKnownFor` has the same issue as below. Just `for` suffices.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Tracer.swift [hunk 8]
```diff
@@ -121,7 +121,8 @@ extension ModuleDependencyGraph.Tracer {
         path.compactMap { node in
           node.dependencySource.map {
             source in
-            graph.inputDependencySourceMap[source].map { input in
+            graph.inputDependencySourceMap.input(ifKnownFor: source).map {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Tracer.swift:124]: Any reason this is here? Seems superfluous.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /// Add a mapping to & from input source file to swiftDeps file path
+  ///
+  /// - Parameter input: the source file path
+  /// - Parameter dependencySource: the dependency source (holding the swiftdeps path)
+  /// - Parameter why: the purpose for this addition. Will be used for future restructuring.
+  @_spi(Testing) public mutating func addEntry(_ input: TypedVirtualPath,
+                                               _ dependencySource: DependencySource,
+                                               for why: AdditionPurpose) {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:89]: It looks like the `for` parameter is dead. Is that intentional?

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 10]
```diff
@@ -214,17 +200,47 @@ extension ModuleDependencyGraph {
...
+      else {
+        // Don't break in order to report all failures.
+        allFound = false
+      }
+    }
+    return allFound
+  }
+}
+extension OutputFileMap {
+  fileprivate func getDependencySource(
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:224]: Nit from before: `get` prefixes usually imply indirect return values in Cocoa and Swift. Just `dependencySource(for:diagnosticEngine:)` suffices.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 11]
```diff
@@ -214,17 +200,47 @@ extension ModuleDependencyGraph {
...
-    return info.inputFiles.reduce(true) { okSoFar, input in
-      ofm.getDependencySource(for: input, diagnosticEngine: de)
-        .map {source in addMapEntry(input, source); return okSoFar } ?? false
+    let diags = info.diagnosticEngine
+    var allFound = true
+    for input in info.inputFiles {
+      if let source = ofm.getDependencySource(for: input, diagnosticEngine: diags) {
+        inputDependencySourceMap.addEntry(input, source, for: purpose)
+      }
+      else {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:215]: Nit: OTBS here please

```suggestion
      if let source = ofm.getDependencySource(for: input, diagnosticEngine: diags) {
         inputDependencySourceMap.addEntry(input, source, for: purpose)
       } else {
```

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 12]
```diff
@@ -66,22 +54,20 @@ import SwiftOptions
...
-    assert(input.type == .swift && dependencySource.typedFile.type == .swiftDeps)
-    inputDependencySourceMap[input] = dependencySource
-  }
-
-  @_spi(Testing) public func getSource(for input: TypedVirtualPath,
-                                       function: String = #function,
-                                       file: String = #file,
-                                       line: Int = #line) -> DependencySource {
-    guard let source = inputDependencySourceMap[input] else {
+  @_spi(Testing) public func source(requiredFor input: TypedVirtualPath,
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:57]: The `requirement` part of this API is communicated by the type signature being non-optional. You could reinforce this in documentation, but I think `for` suffices.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 13]
```diff
@@ -329,7 +345,8 @@ extension ModuleDependencyGraph {
   ) -> TransitivelyInvalidatedInputSet? {
     var invalidatedInputs = TransitivelyInvalidatedInputSet()
     for invalidatedSwiftDeps in collectSwiftDepsUsingInvalidated(nodes: directlyInvalidatedNodes) {
-      guard let invalidatedInput = getInput(for: invalidatedSwiftDeps) else {
+      guard let invalidatedInput = input(neededFor: invalidatedSwiftDeps)
+      else {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:349]: Put this back if you could. This fits onto 80 columns just fine.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift (hunk 10 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:224]: Thanks--done!

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift (hunk 11 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:215]: I did a look-up on "OTBS" and it came back with "off track horse betting"! :)

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:215]: Looks like the request is to put the else on the same line as the closing brace--no prob!

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift (hunk 13 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:349]: What's the rule you follow in general for the "else" in a guard? 80-columns?

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:349]: Will do.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift (hunk 12 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:57]: By putting the requirement in the function name, it becomes more salient at the point of use, *without* clicking to see the documentation, or the type of the assigned-to variable.
Consider: `let foo = source(requiredFor: someInput)` vs `let foo = source(for: someInput)`.
The first construction makes the required nature of the mapping more salient.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 9 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:89]: Yes. As the documentation says, it will be used in the future. For now, it makes clear at the point-of-use what the purpose is.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Tracer.swift (hunk 8 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Tracer.swift:124]: The "ifKnownFor" is there because it makes two aspects more salient:
1. That the function might not return a value. (In this use, it reinforces that the `map` is to handle an optional, rather than a collection.), and
2. That it depends on if the mapping is "known". I'd be happy to find a better word than "known". "Has been seen" feels too wordy.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift (hunk 7 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/InputDependencySourceMap.swift:51]: Please see my comment for the next issue. I have a rule: Correlate salience with importance in order to ease the comprehension for new maintainer.

**Comment by davidungar** [comment]
@CodaFi Thank you for your review. I have responded to your comments and made changes where I agree. You have a wealth of experience with Swift style, and that helps my code fit in. To clarify, I think we agree that this PR makes no functional change, and that (for me) is the most critical part of your review, because I don't want to break the driver. There are some remaining disagreements on function naming, but I think they follow from my practice of choosing salience at the point-of-use. If Xcode were better at showing inferred types, or Doxygen documentation, I'd be in agreement with you. 

Would you be able to approve this PR at this point?

**Comment by davidungar** [comment]
@swift-ci please test

**Review by CodaFi — APPROVED** [review]
I remain unconvinced that this is a meaningful change, and moreover that the new API surface here is an improvement over the status quo. But this PR has also been blocked on my review and I don’t see a reason for that much, so

**Comment by davidungar** [comment]
> I remain unconvinced that this is a meaningful change, and moreover that the new API surface here is an improvement over the status quo. But this PR has also been blocked on my review and I don’t see a reason for that much, so

I understand. Thank you so very much for compromising here.
